# Ejercicios: Parte I

<div class="alert-warning">
    
Indicaciones: Los ejercicios 1, 2, 3 y 4 deben leer su data de links en Github y el código y los resultados deben ser publicados en páginas en Github.

</div>

## Ejercicio 1

<div class="alert-success">
    
1. Use two maps of points.

2. Compute the distance matrix for both maps.

3. Select one row of the distance matrix, and plot the two points with the minimal distance on top of the country of your choosing.
</div>

**1.1. Mapas de puntos a emplear:** 
- Plantas de energía solar en China (China3415_SolarPowerPlants)
- Plantas de energía eólica en China (China3415_WindPowerPlants)

- Ambos se encuentran proyectados y presentan un CRS de 3415.

Fuente empleada: [Global Power Plant Database](https://datasets.wri.org/dataset/globalpowerplantdatabase). Estos datos fueron preprocesados en otro Jupyter Notebook.


In [94]:
import geopandas as gpd
import os

China3415_PowerPlants = "https://github.com/Ciencia-de-datos-espaciales-2023-2/Evaluacion4_1ra-Parte_Mining_GeoDFs/raw/main/maps/China_3415/China3415_PowerPlant.gpkg"

China3415_SolarPowerPlants = gpd.read_file(China3415_PowerPlants,layer="Solar")
China3415_SolarPowerPlants.head()

SolarPowerPlant_Name   gppd_idnr  capacity_mw  latitude  longitude  \
0              Aksai A  WKS0070352         20.0    38.969     94.421   
1               Aksu 1  WKS0065982         40.0    41.048     80.123   
2               Aksu A  WKS0065981         40.0    41.045     80.106   
3               Aksu B  WKS0065980         10.0    41.050     80.108   
4               Aksu C  WKS0065979         15.0    41.052     80.110   

  primary_fuel                          geometry  
0        Solar  POINT (-1280489.125 2632839.774)  
1        Solar  POINT (-2513925.820 3088728.626)  
2        Solar  POINT (-2515490.276 3088702.084)  
3        Solar  POINT (-2515190.400 3089243.087)  
4        Solar  POINT (-2514965.303 3089436.855)

In [95]:
China3415_WindPowerPlants = gpd.read_file(China3415_PowerPlants,layer="Wind")
China3415_WindPowerPlants.head()

WindPowerPlant_Name   gppd_idnr  capacity_mw  \
0             Abag Banner Huiteng Liang Phase 1  WRI1072152         49.0   
1                      Anhui Laian Baoshan Wind  WRI1071797         49.0   
2                  Anhui Laian Longtougang Wind  WRI1071802         49.0   
3                     Anhui Laian Longwosi Wind  WRI1071766         49.0   
4  Anning River Canyon (Phase 1) Dechang County  WRI1071914         16.0   

   latitude  longitude primary_fuel                         geometry  
0   43.3500   115.9000         Wind   POINT (667016.352 3043941.695)  
1   32.6739   118.4318         Wind   POINT (923995.020 1806771.074)  
2   32.6010   118.5350         Wind   POINT (934100.135 1798804.909)  
3   32.6700   118.4500         Wind   POINT (925748.157 1806378.403)  
4   27.1280   102.9770         Wind  POINT (-596812.829 1216949.254)

**1.2. Matriz de distancias entre ambos mapas**

Nota: Ambos mapas presentan proyecciones en metros. Pero, la matriz de distancias se encuentra en unidades de kilómetros y recopila distancias entre una planta de energía eólica a una de energía solar.

In [23]:
China3415_SolarPowerPlants_n = China3415_SolarPowerPlants.set_index("SolarPowerPlant_Name")
China3415_WindPowerPlants_n = China3415_WindPowerPlants.set_index("WindPowerPlant_Name")

DistanceMatrix_WindSolarPlants = China3415_WindPowerPlants_n.geometry.apply\
(lambda g: China3415_SolarPowerPlants_n.geometry.distance(g)/1000)

# Visualización de las 10 primeras filas de la matriz de distancias:
DistanceMatrix_WindSolarPlants.head(10)

SolarPowerPlant_Name                              Aksai A       Aksu 1  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             1990.422662  3181.257451   
Anhui Laian Baoshan Wind                      2354.174981  3669.157241   
Anhui Laian Longtougang Wind                  2366.436085  3681.410896   
Anhui Laian Longwosi Wind                     2355.954444  3670.937096   
Anning River Canyon (Phase 1) Dechang County  1572.310161  2679.343247   
Ansai Wangjiawan                              1357.710763  2667.121472   
Baibuluo                                      1819.675330  3062.502867   
Baishi                                        2421.628117  3664.830821   
Baiyun Ebo Inner Mongolia Luneng Phase 2      1436.130154  2664.409583   
Baiyunebo                                     1438.212578  2665.704037   

SolarPowerPlant_Name                               Aksu A       Aksu B  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             3182.821379  3182.529188   
Anhui Laian Baoshan Wind                      3670.613874  3670.521873   
Anhui Laian Longtougang Wind                  3682.866917  3682.775569   
Anhui Laian Longwosi Wind                     3672.393767  3672.301724   
Anning River Canyon (Phase 1) Dechang County  2680.444330  2680.607525   
Ansai Wangjiawan                              2668.613826  2668.480146   
Baibuluo                                      3064.056767  3063.812720   
Baishi                                        3666.387116  3666.135196   
Baiyun Ebo Inner Mongolia Luneng Phase 2      2665.965208  2665.715599   
Baiyunebo                                     2667.259879  2667.009536   

SolarPowerPlant_Name                               Aksu C       Aksu D  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             3182.306878  3182.371500   
Anhui Laian Baoshan Wind                      3670.378673  3670.484468   
Anhui Laian Longtougang Wind                  3682.632648  3682.738553   
Anhui Laian Longwosi Wind                     3672.158508  3672.264295   
Anning River Canyon (Phase 1) Dechang County  2680.581790  2680.720694   
Ansai Wangjiawan                              2668.319471  2668.417885   
Baibuluo                                      3063.608241  3063.684164   
Baishi                                        3665.927739  3666.001878   
Baiyun Ebo Inner Mongolia Luneng Phase 2      2665.509020  2665.583692   
Baiyunebo                                     2666.802681  2666.877186   

SolarPowerPlant_Name                               Aksu E       Aksu F  \
WindPowerPlant_Name                                                      
Abag Banner Huiteng Liang Phase 1             3183.209002  3184.188716   
Anhui Laian Baoshan Wind                      3671.248527  3672.521166   
Anhui Laian Longtougang Wind                  3683.502237  3684.775572   
Anhui Laian Longwosi Wind                     3673.028377  3674.300973   
Anning River Canyon (Phase 1) Dechang County  2681.277660  2682.705241   
Ansai Wangjiawan                              2669.204159  2670.429227   
Baibuluo                                      3064.512181  3065.578118   
Baishi                                        3666.831844  3667.884389   
Baiyun Ebo Inner Mongolia Luneng Phase 2      2666.413094  2667.469681   
Baiyunebo                                     2667.706766  2668.762103   

SolarPowerPlant_Name                           Aksu Jinko     Ala'er A  ...  \
WindPowerPlant_Name                                                     ...   
Abag Banner Huiteng Liang Phase 1             3181.660867  3086.424381  ...   
Anhui Laian Baoshan Wind                      3669.320824  3559.211154  ...   
Anhui Laian Longtougang Wind                  3681.573685  3571.437685  ...   
Anhui Laian Longwosi Wind                     367

**1.3. Seleccionar una fila de la matriz de distancias y graficas dos puntos con la mínima distancia sobre el mapa del país.**

Se ha seleccionado la fila correspondiente con la planta de energía eólica "Abag Banner Huiteng Liang Phase 1", se determinó a cuál planta de energía solar se encontraba más cerca y ambos puntos se graficaron en un mapa interactivo de China.

In [121]:
# Planta de energía solar más cercana a la planta eólica "Abag Banner Huiteng Liang Phase 1"
AbagBanner_ClosestSolarPlant = DistanceMatrix_WindSolarPlants.loc["Abag Banner Huiteng Liang Phase 1"].idxmin(axis=0)

#Ambos puntos:
Abag_Banner = China3415_WindPowerPlants[
    China3415_WindPowerPlants.WindPowerPlant_Name=="Abag Banner Huiteng Liang Phase 1"]

AbagBanner_ClosestSolarPlant_p = China3415_SolarPowerPlants[
    China3415_SolarPowerPlants.SolarPowerPlant_Name == AbagBanner_ClosestSolarPlant]

#Mapa interactivo con ambos puntos 
base = Abag_Banner.explore(color="blue",zoom_start =3.5, marker_kwds=dict(radius=3))
AbagBanner_ClosestSolarPlant_p.explore(m=base, color="red",marker_kwds=dict(radius=3))

## Ejercicio 2

<div class="alert-success">
    
1. Use a map of points and a map of lines.

2. Compute the distance matrix for both maps.

3. Select one line of the distance matrix, and plot the closests and the farthest point to that line.
   
    
</div>

**2.1. Mapas de puntos y líneas a emplear:**
- Mapa de puntos: Plantas de energía nuclear en China (China3415_NuclearPowerPlants).

- Mapa de líneas: Arroyos en China (China3415_Streams).

Fuentes empleadas: [Global Power Plant Database](https://datasets.wri.org/dataset/globalpowerplantdatabase), [Stream Network dataset](https://data.apps.fao.org/catalog/iso/dc2a5121-0b32-482b-bd9b-64f7a414fa0d). 

In [97]:
China3415_NuclearPowerPlants = gpd.read_file(China3415_PowerPlants,layer="Nuclear")
China3415_NuclearPowerPlants.head()

NuclearPowerPlant_Name   gppd_idnr  capacity_mw  latitude  longitude  \
0                   CEFR  WRI1061016         20.0     39.74     116.03   
1             Changjiang  WRI1061017       1220.0     19.46     108.90   
2               Daya Bay  WRI1061018       1888.0     22.60     114.54   
3          Fangchenggang  WRI1061019       2000.0     21.67     108.56   
4            Fangjiashan  WRI1061020       2024.0     30.44     120.94   

  primary_fuel                         geometry  
0      Nuclear   POINT (683894.146 2615056.503)  
1      Nuclear    POINT (-34893.586 338256.930)  
2      Nuclear    POINT (555446.257 677037.005)  
3      Nuclear    POINT (-62221.064 583652.694)  
4      Nuclear  POINT (1174773.275 1563835.934)

In [98]:
China3415_GeoMaps = "https://github.com/Ciencia-de-datos-espaciales-2023-2/Evaluacion4_1ra-Parte_Mining_GeoDFs/raw/main/maps/China_3415/China3415_GeoMap.gpkg"
China3415_Streams = gpd.read_file(China3415_GeoMaps,layer="Streams") 
China3415_Streams.head()

Stream_UniqueArcIdentifier_ARCID Major_Basin          Sub_Basin  \
0                            147498      Hainan  Hainan East Coast   
1                            147474      Hainan  Hainan East Coast   
2                            147459      Hainan  Hainan East Coast   
3                            147485      Hainan  Hainan East Coast   
4                            147735      Hainan  Hainan East Coast   

                                            geometry  
0  LINESTRING (51764.012 218191.395, 52332.576 21...  
1  LINESTRING (57538.001 220114.352, 57502.826 21...  
2  LINESTRING (76269.475 221017.047, 73797.678 21...  
3  LINESTRING (39797.039 223356.095, 40437.957 22...  
4  LINESTRING (31511.974 202367.728, 31492.798 20...

**2.2. Matriz de distancias entre ambos mapas**

Nota: Ambos mapas presentan proyecciones en metros. Pero, la matriz de distancias se encuentra en unidades de kilómetros.

In [56]:
China3415_NuclearPowerPlants_n = China3415_NuclearPowerPlants.set_index("NuclearPowerPlant_Name")
China3415_Streams_n = China3415_Streams.set_index("Stream_UniqueArcIdentifier_ARCID")

DistanceMatrix_StreamNuclearPlant = China3415_Streams_n.geometry.apply\
(lambda g: China3415_NuclearPowerPlants_n.geometry.distance(g)/1000)

# Visualización de las 10 primeras filas de la matriz de distancias:
DistanceMatrix_StreamNuclearPlant.head(10)

NuclearPowerPlant_Name                   CEFR  Changjiang    Daya Bay  \
Stream_UniqueArcIdentifier_ARCID                                        
147498                            2478.820456  148.071847  681.348000   
147474                            2475.493876  150.004222  675.789126   
147459                            2469.945882  161.353787  661.486652   
147485                            2476.286481  137.043392  681.026276   
147735                            2499.333836  151.246731  706.978115   
147727                            2494.731976  144.813159  705.564204   
147696                            2494.795582  140.916364  708.962731   
147323                            2477.662409  111.183609  709.138990   
147344                            2466.460984  111.881672  687.619169   
146955                            2443.286224   90.641712  668.181898   

NuclearPowerPlant_Name            Fangchenggang  Fangjiashan       Fuqing  \
Stream_UniqueArcIdentifier_ARCID                                            
147498                               382.824449  1752.686233  1267.321384   
147474                               382.756267  1747.513190  1261.653503   
147459                               388.180682  1734.898603  1246.431720   
147485                               374.461390  1751.796322  1267.266474   
147735                               392.637375  1777.819473  1293.080905   
147727                               386.429511  1775.548444  1292.098999   
147696                               383.576693  1778.185616  1295.817452   
147323                               356.295188  1773.588480  1297.540409   
147344                               355.866104  1754.377970  1275.526608   
146955                               329.428561  1733.542093  1256.603495   

NuclearPowerPlant_Name              Hongyanhe     Ling Ao       Ningde  \
Stream_UniqueArcIdentifier_ARCID                                         
147498                            2667.572830  682.110064  1443.381637   
147474                            2663.397237  676.548675  1437.805301   
147459                            2654.756399  662.233545  1423.329896   
147485                            2665.663959  681.791246  1443.081926   
147735                            2690.488389  707.742054  1469.027512   
147727                            2686.746551  706.334222  1467.636793   
147696                            2687.876636  709.737600  1471.032864   
147323                            2675.483724  709.940669  1470.836641   
147344                            2660.767916  688.410864  1449.532133   
146955                            2639.073894  668.983148  1429.902256   

NuclearPowerPlant_Name              Qinshan 2      Tianwan   Yangjiang  
Stream_UniqueArcIdentifier_ARCID                                        
147498                            1752.767478  2059.427691  450.084480  
147474                            1747.592137  2055.032856  445.111214  
147459                            1734.967546  2045.575079  433.575587  
147485                            1751.880487  2057.710958  448.979210  
147735                            1777.904532  2082.909816  474.912826  
147727                            1775.637763  2079.445611  472.398714  
147696                            1778.278822  2080.889616  474.860079  
147323                            1773.703185  2070.054277  470.147376  
147344                            1754.481256  2054.324349  450.819446  
146955                            1733.650115  2032.891635  429.983008

**2.3. Seleccionar una fila de la matriz de distancias y graficar el punto/línea más cercano y lejano.**

Se ha seleccionado la fila correspondiente al arroyo con identificador ARCID 146955, se determinó a cuál planta de energía nuclear se encontraba más cerca y lejos. Los dos puntos y la línea se graficaron en un mapa interactivo de China.

Nota: El arroyo se ve como una línea muy pequeña, porque su masa de agua es menor comparada con la de un río.

In [120]:
# Planta de energía nuclear más cercana y lejana al arroyo con identificador ARCID 146955
Stream146955_ClosestNuclearPlant = DistanceMatrix_StreamNuclearPlant.loc[146955].idxmin(axis=0)
Stream146955_FurthestNuclearPlant = DistanceMatrix_StreamNuclearPlant.loc[146955].idxmax(axis=0)

#Coordenadas de ambos puntos y el arroyo con identificador ARCID 146955
Stream146955_ClosestNuclearPlant_p = China3415_NuclearPowerPlants[
    China3415_NuclearPowerPlants.NuclearPowerPlant_Name==Stream146955_ClosestNuclearPlant]

Stream146955_FurthestNuclearPlant_p = China3415_NuclearPowerPlants[
    China3415_NuclearPowerPlants.NuclearPowerPlant_Name==Stream146955_FurthestNuclearPlant]

Stream146955_p = China3415_Streams[China3415_Streams.Stream_UniqueArcIdentifier_ARCID==146955]

#Mapa interactivo con ambos puntos y el arroyo con identificador ARCID 146955
base = Stream146955_p.explore(color="blue",zoom_start=4)
Stream146955_ClosestNuclearPlant_p.explore(m=base, color="red",marker_kwds=dict(radius=3))
Stream146955_FurthestNuclearPlant_p.explore(m=base, color="green",marker_kwds=dict(radius=3))

## Ejercicio 3

<div class="alert-success">
    
1. Create a set of points and a set of polygons

2. Compute the distance matrix for both sets.

3. Select one polygon of the distance matrix, and plot the closests and the farthest point to that polygon.
    
</div>   

**3.1. Set de puntos y polígonos a emplear:**
- Set de puntos: Plantas de energía nuclear en China (China3415_NuclearPowerPlants).Es el mismo que se empleó en la pregunta 2.

- Set de polígonos: Convex Hulls de arroyos en China que pertenecen a una misma cuenca principal (China3415_MajorStreamBasin_Hulls). El GeoDataFrame con los convex hulls ya fue creado en el Jupyter Notebook de preprocesamiento.


In [122]:
China3415_MajorStream_BasinsHulls = gpd.read_file(China3415_GeoMaps,layer="MajorStream_BasinsHulls") 
China3415_MajorStream_BasinsHulls.head()

# Notar que "Major_Basin" alude a la cuenca principal donde se ha agrupado un conjunto de arroyos

Major_Basin  \
0                              Amur   
1  Bo Hai - Korean Bay, North Coast   
2                       China Coast   
3               Ganges - Bramaputra   
4                     Gobi Interior   

                                            geometry  
0  POLYGON ((1714301.771 2910473.532, 1537547.570...  
1  POLYGON ((1230941.409 2559984.625, 963188.294 ...  
2  POLYGON ((505330.365 666112.023, 503617.109 66...  
3  POLYGON ((-1674523.275 1396331.184, -1979103.4...  
4  POLYGON ((-519694.829 2358194.060, -569682.801...

**3.2. Matriz de distancias entre ambos mapas**

Nota: La matriz de distancias se encuentra en unidades de kilómetros.

In [91]:
China3415_MajorStream_BasinsHulls_n = China3415_MajorStream_BasinsHulls.set_index("Major_Basin")

DistanceMatrix_StreamBasinsHulls_NuclearPlant = China3415_MajorStream_BasinsHulls_n.geometry.apply\
(lambda g: China3415_NuclearPowerPlants_n.geometry.distance(g)/1000)

# Visualización de las 10 primeras filas de la matriz de distancias:
DistanceMatrix_StreamBasinsHulls_NuclearPlant.head(10)

NuclearPowerPlant_Name                   CEFR   Changjiang     Daya Bay  \
Major_Basin                                                               
Amur                               679.388929  3009.594588  2438.981415   
Bo Hai - Korean Bay, North Coast   228.475546  2472.452237  1966.415288   
China Coast                        392.981428   620.417018     0.000000   
Ganges - Bramaputra               2120.404089  1532.320412  1832.754215   
Gobi Interior                      269.736509  2070.761509  1936.420574   
Hainan                            2279.079778     0.000000   485.064388   
Hong (Red River)                  2087.226470   501.962473   884.870624   
Huang He                            32.350788  1532.956299  1277.793648   
Indus                             3314.756499  2951.754026  3334.118214   
Irrawaddy                         2122.313552  1191.575245  1630.629683   

NuclearPowerPlant_Name            Fangchenggang  Fangjiashan       Fuqing  \
Major_Basin                                                                 
Amur                                2811.009362  1397.413050  1973.975974   
Bo Hai - Korean Bay, North Coast    2262.261451   997.730963  1569.474959   
China Coast                          533.166772     0.000000    16.651114   
Ganges - Bramaputra                 1351.292290  2284.688847  2201.308181   
Gobi Interior                       1825.379265  1429.137229  1872.212425   
Hainan                               222.677882  1547.803213  1074.682368   
Hong (Red River)                     310.681145  1702.877603  1398.327721   
Huang He                            1286.086765   681.932724  1148.411409   
Indus                               2807.811105  3745.025648  3709.382150   
Irrawaddy                           1060.911928  2228.078841  2089.632823   

NuclearPowerPlant_Name              Hongyanhe      Ling Ao       Ningde  \
Major_Basin                                                               
Amur                               413.208049  2438.573933  1780.045817   
Bo Hai - Korean Bay, North Coast     0.000000  1966.199124  1381.830381   
China Coast                        238.812795     0.000000     0.000000   
Ganges - Bramaputra               2564.211323  1833.684551  2248.927315   
Gobi Interior                      564.656228  1936.863642  1746.612206   
Hainan                            2460.864248   485.896753  1245.817608   
Hong (Red River)                  2400.866379   885.894373  1514.396404   
Huang He                           323.278011  1278.008157  1011.743087   
Indus                             3805.981091  3335.076096  3747.549667   
Irrawaddy                         2553.011558  1631.638578  2166.516834   

NuclearPowerPlant_Name              Qinshan 2      Tianwan    Yangjiang  
Major_Basin                                                              
Amur                              1397.885615  1012.444800  2627.316040  
Bo Hai - Korean Bay, North Coast   998.415499   547.874727  2120.878176  
China Coast                          0.000000     0.000000   191.334274  
Ganges - Bramaputra               2285.665828  2202.074934  1670.965223  
Gobi Interior                     1430.261066   933.420035  1938.894073  
Hainan                            1547.911195  1851.860761   244.248103  
Hong (Red River)                  1703.428669  1850.593980   668.115234  
Huang He                           683.068163   201.889659  1339.639315  
Indus                             3746.098718  3576.113229  3154.243865  
Irrawaddy                         2229.028730  2163.456066  1424.257904

**3.3. Seleccionar un polígono y graficar el punto más cercano y lejano.**

Se ha seleccionado la fila correspondiente al polígono de la cuenca "Huang He", se determinó a cuál planta de energía nuclear se encontraba más cerca y lejos. Los dos puntos y el polígono se graficaron en un mapa interactivo de China.

In [135]:
# Planta de energía nuclear más cercana y lejana al polígono de la cuenca "Huang He"
HuangHeHull_ClosestNuclearPlant = DistanceMatrix_StreamBasinsHulls_NuclearPlant.loc["Huang He"].idxmin(axis=0)
HuangHeHull_FurthestNuclearPlant = DistanceMatrix_StreamBasinsHulls_NuclearPlant.loc["Huang He"].idxmax(axis=0)

#Coordenadas de ambos puntos y el polígono
HuangHeHull_ClosestNuclearPlant_p = China3415_NuclearPowerPlants[
    China3415_NuclearPowerPlants.NuclearPowerPlant_Name==HuangHeHull_ClosestNuclearPlant]

HuangHeHull_FurthestNuclearPlant_p = China3415_NuclearPowerPlants[
    China3415_NuclearPowerPlants.NuclearPowerPlant_Name==HuangHeHull_FurthestNuclearPlant]

HuangHeHull = China3415_MajorStream_BasinsHulls[China3415_MajorStream_BasinsHulls.Major_Basin=="Huang He"]

#Mapa interactivo con ambos puntos y el polígono
base = HuangHeHull.explore(color="blue",zoom_start=4)
HuangHeHull_ClosestNuclearPlant_p.explore(m=base, color="red",marker_kwds=dict(radius=3))
HuangHeHull_FurthestNuclearPlant_p.explore(m=base, color="green",marker_kwds=dict(radius=3))

## Ejercicio 4

<div class="alert-success">
    
1. Create a set of points and a set of lines

2. Get the buffer for the lines, select different values for the distance.

3. Keep the points that are within the buffer (as in point 2, you need to play with differn distances until you show something interesting).
   

</div>   

**4.1. Set de puntos y polígonos a emplear:**
- Set de puntos: Plantas de energía nuclear en China (China3415_NuclearPowerPlants). Es el mismo que se empleó en las preguntas 2 y 3.

- Set de líneas: Conjunto de arroyos en China que pertenecen a una misma cuenca principal (China3415_MajorStreamBasin). El GeoDataFrame con las multilíneas correspondientes a un cuenca principal ya fue creado en el Jupyter Notebook de preprocesamiento.


In [126]:
China3415_MajorStream_Basins = gpd.read_file(China3415_GeoMaps,layer="MajorStream_Basins") 
China3415_MajorStream_Basins.head()

# Notar que "Major_Basin" alude a la cuenca principal donde se ha agrupado un conjunto de arroyos

Major_Basin  \
0                              Amur   
1  Bo Hai - Korean Bay, North Coast   
2                       China Coast   
3               Ganges - Bramaputra   
4                     Gobi Interior   

                                            geometry  
0  MULTILINESTRING ((1786928.569 2972264.304, 178...  
1  MULTILINESTRING ((1420683.342 2663487.794, 142...  
2  MULTILINESTRING ((539323.458 689436.814, 53996...  
3  MULTILINESTRING ((-1271163.883 1640257.720, -1...  
4  MULTILINESTRING ((-519694.829 2358194.060, -52...

**4.2. *Buffers* alrededor de la cuenca China Coast**

Se crearon *buffers* de los arroyos que forman parte de la cuenca "China Coast". Se seleccionó esta cuenca, porque se localiza al sureste de China, donde la mayoría de las plantas nucleares también se encuentran.

Primero, con la finalidad de evaluar qué rango de distancias emplear para crear los *buffers*, se creó una matriz de distancias:

In [181]:
China3415_MajorStream_Basins_n = China3415_MajorStream_Basins.set_index("Major_Basin")
DistanceMatrix_StreamBasins_NuclearPlant = China3415_MajorStream_Basins_n.geometry.apply\
(lambda g: China3415_NuclearPowerPlants_n.geometry.distance(g)/1000)

import pandas as pd
ChinaCoast_Distances = pd.Series(list(DistanceMatrix_StreamBasins_NuclearPlant.loc["China Coast"]))
ChinaCoast_Distances.describe()

count     12.000000
mean     173.754462
std      227.193252
min        4.340807
25%       12.163869
50%       20.341937
75%      280.885002
max      620.417018
dtype: float64

Con los resultados obtenidos de esta evaluación preliminar, se observa que la planta nuclear más cercana a la cuenca "Coast China" está a 4.34 km, mientras que la más lejana, a 620.42 km. Con esto en consideración, se crearán buffers de 5 km y 10 km. No fue posible trabajar con mayores distancias, porque el tiempo de computación requerido era elevado.

In [ ]:
ChinaCoastBuffer_5km = China3415_MajorStream_Basins[
    China3415_MajorStream_Basins.Major_Basin=="China Coast"
        ].buffer(distance = 5*1000)

ChinaCoastBuffer_5km 

**4.3. Mantener los puntos que se encuentran dentro de cada *buffer* y verificar con cuál se obtienen los resultados más interesantes**

In [ ]:
NuclearPowerPlants_WithinBuffer_5km = China3415_NuclearPowerPlants.clip(mask=ChinaCoastBuffer_5km)
NuclearPowerPlants_WithinBuffer_5km